doc/f2py/f2py.1
.TH "F2PY" 1
.SH NAME
f2py \- Fortran to Python interface generator
.SH SYNOPSIS
(1) To construct extension module sources:

.B f2py
[<options>] <fortran files> [[[only:]||[skip:]] <fortran functions> ] [: <fortran files> ...]

(2) To compile fortran files and build extension modules:

.B f2py
\-c [<options>, <config_fc options>, <extra options>] <fortran files>

(3) To generate signature files:

.B f2py
\-h <filename.pyf> ...< same options as in (1) >
.SH DESCRIPTION
This program generates a Python C/API file (<modulename>module.c)
that contains wrappers for given Fortran or C functions so that they
can be called from Python.
With the \-c option the corresponding
extension modules are built.
.SH OPTIONS
.TP
.B \-h <filename>
Write signatures of the fortran routines to file <filename> and
exit. You can then edit <filename> and use it instead of <fortran
files>. If <filename>==stdout then the signatures are printed to
stdout.
.TP
.B <fortran functions>
Names of fortran routines for which Python C/API functions will be
generated. Default is all that are found in <fortran files>.
.TP
.B skip:
Ignore fortran functions that follow until `:'.
.TP
.B only:
Use only fortran functions that follow until `:'.
.TP
.B :
Get back to <fortran files> mode.
.TP
.B \-m <modulename>
Name of the module; f2py generates a Python/C API file
<modulename>module.c or extension module <modulename>.  Default is
\'untitled\'.
.TP
.B \-\-[no\-]lower
Do [not] lower the cases in <fortran files>. By default, \-\-lower is
assumed with \-h key, and \-\-no\-lower without \-h key.
.TP
.B \-\-build\-dir <dirname>
All f2py generated files are created in <dirname>. Default is tempfile.mkdtemp().
.TP
.B \-\-overwrite\-signature
Overwrite existing signature file.
.TP
.B \-\-[no\-]latex\-doc
Create (or not) <modulename>module.tex.  Default is \-\-no\-latex\-doc.
.TP
.B \-\-short\-latex
Create 'incomplete' LaTeX document (without commands \\documentclass,
\\tableofcontents, and \\begin{document}, \\end{document}).
.TP
.B \-\-[no\-]rest\-doc
Create (or not) <modulename>module.rst.  Default is \-\-no\-rest\-doc.
.TP
.B \-\-debug\-capi
Create C/API code that reports the state of the wrappers during
runtime. Useful for debugging.
.TP
.B \-include\'<includefile>\'
Add CPP #include statement to the C/API code.  <includefile> should be
in the format of either `"filename.ext"' or `<filename.ext>'.  As a
result <includefile> will be included just before wrapper functions
part in the C/API code. The option is depreciated, use `usercode`
statement in signature files instead.
.TP
.B \-\-[no\-]wrap\-functions
Create Fortran subroutine wrappers to Fortran 77
functions. \-\-wrap\-functions is default because it ensures maximum
portability/compiler independence.
.TP
.B \-\-help\-link [..]
List system resources found by system_info.py. [..] may contain
a list of resources names. See also \-\-link\-<resource> switch below.
.TP
.B \-\-quiet
Run quietly.
.TP
.B \-\-verbose
Run with extra verbosity.
.TP
.B \-v
Print f2py version ID and exit.
.TP
.B \-\-include_paths path1:path2:...
Search include files (that f2py will scan) from the given directories.
.SH "CONFIG_FC OPTIONS"
The following options are effective only when \-c switch is used.
.TP
.B \-\-help-compiler
List available Fortran compilers [DEPRECIATED].
.TP
.B \-\-fcompiler=<name>
Specify Fortran compiler type by vendor.
.TP
.B \-\-compiler=<name>
Specify C compiler type (as defined by distutils)
.TP
.B \-\-fcompiler-exec=<path>
Specify the path to F77 compiler [DEPRECIATED].
.TP
.B \-\-f90compiler\-exec=<path>
Specify the path to F90 compiler [DEPRECIATED].
.TP
.B \-\-help\-fcompiler
List available Fortran compilers and exit.
.TP
.B \-\-f77exec=<path>
Specify the path to F77 compiler.
.TP
.B \-\-f90exec=<path>
Specify the path to F90 compiler.
.TP
.B  \-\-f77flags="..."
Specify F77 compiler flags.
.TP
.B \-\-f90flags="..."
Specify F90 compiler flags.
.TP
.B \-\-opt="..."
Specify optimization flags.
.TP
.B \-\-arch="..."
Specify architecture specific optimization flags.
.TP
.B \-\-noopt
Compile without optimization.
.TP
.B \-\-noarch
Compile without arch-dependent optimization.
.TP
.B \-\-debug
Compile with debugging information.
.SH "EXTRA OPTIONS"
The following options are effective only when \-c switch is used.
.TP
.B \-\-link-<resource> 
Link extension module with <resource> as defined by
numpy_distutils/system_info.py. E.g. to link with optimized LAPACK
libraries (vecLib on MacOSX, ATLAS elsewhere), use
\-\-link\-lapack_opt. See also \-\-help\-link switch.

.TP
.B -L/path/to/lib/ -l<libname>
.TP
.B -D<define> -U<name> -I/path/to/include/ 
.TP
.B <filename>.o <filename>.so <filename>.a

.TP
.B -DPREPEND_FORTRAN -DNO_APPEND_FORTRAN -DUPPERCASE_FORTRAN -DUNDERSCORE_G77
Macros that might be required with non-gcc Fortran compilers. 

.TP
.B -DF2PY_REPORT_ATEXIT
To print out a performance report of F2PY interface when python
exits. Available for Linux.

.TP
.B -DF2PY_REPORT_ON_ARRAY_COPY=<int>
To send a message to stderr whenever F2PY interface makes a copy of an
array. Integer <int> sets the threshold for array sizes when a message
should be shown.

.SH REQUIREMENTS
Python 1.5.2 or higher (2.x is supported).

Numerical Python 13 or higher (20.x,21.x,22.x,23.x are supported).

Optional Numarray 0.9 or higher partially supported.

numpy_distutils from Scipy (can be downloaded from F2PY homepage)
.SH "SEE ALSO"
python(1)
.SH BUGS
For instructions on reporting bugs, see 

  http://cens.ioc.ee/projects/f2py2e/FAQ.html
.SH AUTHOR
Pearu Peterson <pearu@cens.ioc.ee>
.SH "INTERNET RESOURCES"
Main website: http://cens.ioc.ee/projects/f2py2e/

User's Guide: http://cens.ioc.ee/projects/f2py2e/usersguide/

Mailing list: http://cens.ioc.ee/mailman/listinfo/f2py-users/

Scipy website: http://www.numpy.org
.SH COPYRIGHT
Copyright (c) 1999, 2000, 2001, 2002, 2003, 2004, 2005 Pearu Peterson
.SH LICENSE
NumPy License
.SH VERSION
2.45.241


numpy/distutils/scons/tools/f2py.py
"""f2py Tool

Tool-specific initialization for f2py.

There normally shouldn't be any need to import this module directly.
It will usually be imported through the generic SCons.Tool.Tool()
selection method.

"""

import os.path
import re

import SCons.Action
import SCons.Defaults
import SCons.Scanner
import SCons.Tool
import SCons.Util
import SCons.Node

# XXX: this whole thing needs cleaning !

def _f2pySuffixEmitter(env, source):
    return '$F2PYCFILESUFFIX'

#_reModule = re.compile(r'%module\s+(.+)')

def _mangle_fortranobject(targetname, filename):
    basename = os.path.splitext(os.path.basename(targetname))[0]
    return '%s_%s' % (basename, filename)
    
def _f2pyEmitter(target, source, env):
    build_dir = os.path.dirname(str(target[0]))
    target.append(SCons.Node.FS.default_fs.Entry(
        os.path.join(build_dir, _mangle_fortranobject(str(target[0]), 'fortranobject.c'))))
    if _is_pyf(str(source[0])):
        basename = os.path.splitext(os.path.basename(str(target[0])))
        basename = basename[0]
        basename = basename.split('module')[0]
        target.append(SCons.Node.FS.default_fs.Entry(
            os.path.join(build_dir, '%s-f2pywrappers.f' % basename)))
    return (target, source)

def _is_pyf(source_file):
    return os.path.splitext(source_file)[1] == '.pyf'

def _pyf2c(target, source, env):
    import numpy.f2py
    import shutil

    # We need filenames from source/target for path handling
    target_file_names = [str(i) for i in target]
    source_file_names = [str(i) for i in source]

    # Get source files necessary for f2py generated modules
    d = os.path.dirname(numpy.f2py.__file__)
    source_c = os.path.join(d,'src','fortranobject.c')

    # Copy source files for f2py generated modules in the build dir
    build_dir = os.path.dirname(target_file_names[0])

    # XXX: blah
    if build_dir == '':
        build_dir = '.'

    try:
        shutil.copy(source_c, os.path.join(build_dir, 
               _mangle_fortranobject(target_file_names[0], 'fortranobject.c')))
    except IOError, e:
        msg = "Error while copying fortran source files (error was %s)" % str(e)
        raise IOError(msg)

    basename = os.path.basename(str(target[0]).split('module')[0])

    # XXX: handle F2PYOPTIONS being a string instead of a list
    if _is_pyf(source_file_names[0]):
        # XXX: scons has a way to force buidler to only use one source file
        if len(source_file_names) > 1:
            raise NotImplementedError("FIXME: multiple source files")
        
        wrapper = os.path.join(build_dir, '%s-f2pywrappers.f' % basename)

        cmd = env['F2PYOPTIONS'] + [source_file_names[0], '--build-dir', build_dir]
        st = numpy.f2py.run_main(cmd)

        if not os.path.exists(wrapper):
            f = open(wrapper, 'w')
            f.close()
    else:
        cmd = env['F2PYOPTIONS'] + source_file_names + ['--build-dir', build_dir]
        # fortran files, we need to give the module name
        cmd.extend(['--lower', '-m', basename])
        st = numpy.f2py.run_main(cmd)

    return 0


def generate(env):
    """Add Builders and construction variables for swig to an Environment."""
    import numpy.f2py
    d = os.path.dirname(numpy.f2py.__file__)

    c_file, cxx_file = SCons.Tool.createCFileBuilders(env)

    c_file.suffix['.pyf'] = _f2pySuffixEmitter

    c_file.add_action('.pyf', SCons.Action.Action(_pyf2c))
    c_file.add_emitter('.pyf', _f2pyEmitter)

    env['F2PYOPTIONS']      = SCons.Util.CLVar('')
    env['F2PYBUILDDIR']     = ''
    env['F2PYCFILESUFFIX']  = 'module$CFILESUFFIX'
    env['F2PYINCLUDEDIR']   = os.path.join(d, 'src')

    # XXX: adding a scanner using c_file.add_scanner does not work...
    expr = '(<)include_file=(\S+)>'
    scanner = SCons.Scanner.ClassicCPP("F2PYScan", ".pyf", "F2PYPATH", expr)
    env.Append(SCANNERS = scanner)

    env['BUILDERS']['F2py'] = SCons.Builder.Builder(action = _pyf2c, 
                                                    emitter = _f2pyEmitter,
                                                    suffix = _f2pySuffixEmitter)

_MINC = re.compile(r'<include_file=(\S+)>')                                              
def _pyf_scanner(node, env, path):
    print "================== SCANNING ===================="
    cnt = node.get_contents()
    return _parse(cnt)

def _parse(lines):                                                                        
    """Return list of included files in .pyf from include_file directive."""
    dep = []                                                                             
    for line in lines:                                                                   
        m = _MINC.search(line)                                                           
        if m:                                                                            
            dep.append(m.group(1))                                                       
    return dep  

def exists(env):
    try:
        import numpy.f2py
        st = 1
    except ImportError, e:
        print "Warning : f2py tool not found, error was %s" % e
        st = 0

    return st
